proxy-socket-udp/src/main/java/net/airvantage/proxysocket/udp/ProxyDatagramSocket.java

```diff
@@ -0,0 +1,102 @@
+ */
+public class ProxyDatagramSocket extends DatagramSocket {
+    private static final Logger LOG = Logger.getLogger(ProxyDatagramSocket.class.getName());
+
+    private ProxyAddressCache addressCache;
+    private ProxyProtocolMetricsListener metrics;
+    private Predicate<InetSocketAddress> trustedProxyPredicate;
+
+    public ProxyDatagramSocket() throws SocketException {
+        super();
+    }
+
+    public ProxyDatagramSocket(SocketAddress bindaddr) throws SocketException {
+        super(bindaddr);
+    }
+
+    public ProxyDatagramSocket(int port) throws SocketException {
+        super(port);
+    }
+
+    public ProxyDatagramSocket(int port, java.net.InetAddress laddr) throws SocketException {
+        super(port, laddr);
+    }
+
+    public ProxyDatagramSocket setCache(ProxyAddressCache cache) { this.addressCache = cache; return this; }
+    public ProxyDatagramSocket setMetrics(ProxyProtocolMetricsListener metrics) { this.metrics = metrics; return this; }
+    public ProxyDatagramSocket setTrustedProxy(Predicate<InetSocketAddress> predicate) { this.trustedProxyPredicate = predicate; return this; }
+
+    @Override
+    public void receive(DatagramPacket packet) throws IOException {
+        super.receive(packet);
+        try {
+            InetSocketAddress lbAddress = (InetSocketAddress) packet.getSocketAddress();
+            if (trustedProxyPredicate != null && !trustedProxyPredicate.test(lbAddress)) {
+                // Untrusted source: do not parse, deliver original packet
+                return;
+            }
+
+            ProxyHeader header = ProxyProtocolV2Decoder.parse(packet.getData(), packet.getOffset(), packet.getLength());
+            if (metrics != null) metrics.onHeaderParsed(header);
+            if (header.isLocal()) {
+                // LOCAL: not proxied
+            } else if (header.isProxy() && header.getProtocol() == ProxyHeader.TransportProtocol.DGRAM) {
+                InetSocketAddress realClient = header.getSourceAddress();
+                if (realClient != null && lbAddress != null) {
+                    if (addressCache != null) addressCache.put(realClient, lbAddress);
+                    packet.setSocketAddress(realClient);
+                }
+            }
+            int headerLen = header.getHeaderLength();
+            packet.setData(packet.getData(), packet.getOffset() + headerLen, packet.getLength() - headerLen);
+        } catch (Exception e) {
+            LOG.log(Level.WARNING, "Proxy socket parse error; delivering original packet.", e);
+            if (metrics != null) metrics.onParseError(e);
+        }
+    }
+
+    @Override
+    public void send(DatagramPacket packet) throws IOException {
+        InetSocketAddress client = (InetSocketAddress) packet.getSocketAddress();
+        InetSocketAddress lb = addressCache != null ? addressCache.get(client) : null;
+        if (lb != null) {
+            packet.setSocketAddress(lb);
```

Comment:
If we agree on previous comment, then there is probably null test to remove

Reply:
this check cannot be removed: even with an active addressCache the data might have expired.
The "solution" in that case to send the packet directly to the real IP might not be perfect however